### DISCUSSION TITLE
Wrong version of tkinter

### DISCUSSION THREAD

**Discussion by saulspatz** [Homebrew/discussions#646] [Everyday usage]
I am using python 3.9.1 on Mac OS 10.15.7 (Catalina).  Python was installed with homebrew. When I import tkinter, I get version 8.5.9, and I want version 8.6 instead.

`brew info tcl-tk` gives 

    tcl-tk: stable 8.6.10 (bottled) [keg-only]
    Tool Command Language
    https://www.tcl-lang.org
    /usr/local/Cellar/tcl-tk/8.6.10 (3,022 files, 50.9MB)
      Poured from bottle on 2020-03-10 at 17:57:25
    From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/tcl-tk.rb
    License: TCL
    ==> Dependencies
    Required: openssl@1.1 ✔
    ==> Caveats
    tcl-tk is keg-only, which means it was not symlinked into /usr/local,
    because macOS already provides this software and installing another version in
    parallel can cause all kinds of trouble.
    
    If you need to have tcl-tk first in your PATH run:
      echo 'export PATH="/usr/local/opt/tcl-tk/bin:$PATH"' >> /Users/saul/.bash_profile
    
    For compilers to find tcl-tk you may need to set:
      export LDFLAGS="-L/usr/local/opt/tcl-tk/lib"
      export CPPFLAGS="-I/usr/local/opt/tcl-tk/include"
    
    For pkg-config to find tcl-tk you may need to set:
      export PKG_CONFIG_PATH="/usr/local/opt/tcl-tk/lib/pkgconfig"
    
    ==> Analytics
    install: 362,394 (30 days), 390,184 (90 days), 474,512 (365 days)
    install-on-request: 9,609 (30 days), 17,845 (90 days), 60,715 (365 days)
    build-error: 0 (30 days)

I put the suggested lines in my `bash_profile` but they don't seem to have had any effect.

    >>> import tkinter
    >>> tcl = tkinter.Tcl()
    >>> print(tcl.call("info", "patchlevel"))
    8.5.9

I found a lot of stuff about this issue online, but it all seemed to be related to pyenv, which I don't use, and I couldn't pick out anything other than what's described above that seemed to related to my situation.

What do I need to do?  

**EDIT**

I tried adding the lines to my .bashrc file also, but that had no effect.

**Comment by carlocab** [comment]
What does `python3 -c 'import sys; print(sys.version)'` give you? What about 
```
/usr/local/bin/python3 -c 'import tkinter; tcl = tkinter.Tcl(); print(tcl.call("info", "patchlevel"))'
```
?

**Reply by saulspatz** [reply]
3.9.1 (default, Dec 28 2020, 11:24:06) 
[Clang 12.0.0 (clang-1200.0.32.28)]

**Reply by carlocab** [reply]
Sorry, I think I edited my comment before you had the chance to see it. See above.

**Reply by saulspatz** [reply]
`/usr/local/bin/python3 -c 'import tkinter; tcl = tkinter.Tcl(); print(tcl.call("info", "patchlevel"))'`
gives `8.5.9`

**Reply by carlocab** [reply]
I think you have an outdated version of python. Can you tell me what the output of
```
brew config && brew doctor && brew info python@3.9
```
says?

Then, do
```
brew update && brew upgrade python@3.9
```
and let me know if that helps.

**Reply by saulspatz** [reply]
The output of  `brew config && brew doctor && brew info python@3.9` was
    
    HOMEBREW_VERSION: 2.7.1
    ORIGIN: https://github.com/Homebrew/brew
    HEAD: fdb71a5cb3029da2ef74d7e09c40575bbfb4ec72
    Last commit: 5 weeks ago
    Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
    Core tap HEAD: 30e7e7a3c90aaf3d0688d59d3643621fb417be78
    Core tap last commit: 4 weeks ago
    Core tap branch: master
    HOMEBREW_PREFIX: /usr/local
    HOMEBREW_CASK_OPTS: []
    HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.9Q0LkB3HAw/org.macosforge.xquartz:0
    HOMEBREW_MAKE_JOBS: 4
    Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
    CPU: quad-core 64-bit ivybridge
    Clang: 12.0 build 1200
    Git: 2.30.0 => /usr/local/bin/git
    Curl: 7.64.1 => /usr/bin/curl
    macOS: 10.15.7-x86_64
    CLT: 12.4.0.0.1.1610135815
    Xcode: 12.4
    XQuartz: 2.7.11 => /opt/X11
    Please note that these warnings are just used to help the Homebrew maintainers
    with debugging if you file an issue. If everything you use Homebrew for is
    working fine: please don't worry or file an issue; just ignore this. Thanks!
    
    Warning: Unbrewed header files were found in /usr/local/include.
    If you didn't put them there on purpose they could cause problems when
    building Homebrew formulae, and may need to be deleted.
    
    Unexpected header files:
      /usr/local/include/python3.8/python-igraph/igraphmodule_api.h

**Reply by saulspatz** [reply]
`brew update && brew upgrade python@3.9` seems to have fixed it.  Thank you so much.

**Comment by iMaupin** [comment]
tkinter is also outdated on python 3.10. Running `brew update && brew upgrade python@3.10`

returns `Warning: python@3.10 3.10.0_2 already installed`

yet my TKVersion still prints as 8.5.

Any thoughts?

**Reply by carlocab** [reply]
What is the output of `brew config`, `brew doctor`, and `brew info python-tk@3.10`?

**Reply by iMaupin** [reply]
brew config:
```
HOMEBREW_VERSION: 3.3.7
ORIGIN: https://github.com/Homebrew/brew
HEAD: dbec9de77ebc65a4845a5e672950219aa19b35b8
Last commit: 28 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: d99ad28fbf75f643c21444e4d728f063d5f664aa
Core tap last commit: 43 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: dodeca-core 64-bit kabylake
Clang: 13.0.0 build 1300
Git: 2.30.1 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.6-x86_64
CLT: 12.5.0.22.9
Xcode: 13.1
```

brew doctor:
```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 13.1.


Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /Users/my_user_name/.pyenv/shims/python3.10-config
  /Users/my_user_name/.pyenv/shims/python-config
  /Users/my_user_name/.pyenv/shims/python3-config

Warning: Broken symlinks were found. Remove them with `brew cleanup`:
  /usr/local/share/man/man5/package-locks.5
  /usr/local/share/man/man5/shrinkwrap-json.5

```

brew info python-tk@3.10:
```
python-tk@3.10: stable 3.10.0 (bottled) [keg-only]
Python interface to Tcl/Tk
https://www.python.org/
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/python-tk@3.10.rb
License: Python-2.0
==> Dependencies
Required: python@3.10 ✔, tcl-tk ✘
==> Caveats
python-tk@3.10 is keg-only, which means it was not symlinked into /usr/local,
because this is an alternate version of another formula.

==> Analytics
install: 419 (30 days), 1,010 (90 days), 1,012 (365 days)
install-on-request: 418 (30 days), 1,008 (90 days), 1,010 (365 days)
build-error: 0 (30 days)
```

**Reply by SMillerDev** [reply]
And is that the python you're actually using? Because you do have a pyenv enabled.

**Reply by iMaupin** [reply]
@SMillerDev yeah good call. I installed 3.10 onto pyenv with the `pyenv install`

**Reply by carlocab** [reply]
If you want to use Homebrew Python 3.10 with tkinter then you need to install `python-tk@3.10`. If you want to use the Python 3.10 you installed with `pyenv` then you need to configure it to use a newer `tcl-tk`, whether installed with Homebrew or otherwise. You'll need to consult the `pyenv` docs on how to do this.

Note that you don't have Homebrew `tcl-tk` installed.

**Comment by iMaupin** [comment]
@carlocab cool. Thank you for the assistance.

**Reply by iMaupin** [reply]
this worked for me.
https://stackoverflow.com/questions/60469202/unable-to-install-tkinter-with-pyenv-pythons-on-macos
